reflection: improve server implementation

jhump opened the pull request:
This updates the server reflection implementation in two key ways:
1. Support alternate source of descriptors, like for RPC servers that get their descriptors dynamically and are dynamic proxies.
    * Instead of always getting descriptors from the set of global registered files (when generated code is linked into the calling program), this allows caller to supply custom resolvers.
    * This also allows caller to supply a set of service names instead of supplying a `*grpc.Server`.
2. Use the new protobuf API v2 stuff to get the descriptors, which is much more sane than the old APIs.
    * This allows much of the code to be eliminated since standard APIs now support symbol and descriptor lookup.
    * The potential downside of this simplification is that encoded descriptors (pre-serialized to bytes) are no longer retained, so descriptor protos are re-created and marshaled with each call. Given the expected frequency of reflection requests, this is likely a non-issue and not worth adding back some sort of cache.
    
RELEASE NOTES:
* A new `NewServer(ServerOptions)` function has been added to the `google.golang.org/grpc/reflection` package that allows creating the reflection server with customizations to the list of advertised services and to the mechanism used to lookup descriptors and extensions.

jhump commented:
With improvements like this, I could also incorporate my custom _wrapper_ around the existing proto registries, like to augment the descriptors with source code info:
https://github.com/jhump/protoreflect/blob/master/desc/sourceinfo/registry.go#L37
Without a change like this, I have to fork the entire implementation of the reflection service.

But, as mentioned in the description, another use case would be a dynamic server/proxy, whose service descriptors might be retrieved dynamically.

jhump commented on reflection/serverreflection_test.go:
```diff
@@ -348,7 +356,7 @@ func testFileContainingSymbol(t *testing.T, stream rpb.ServerReflection_ServerRe
 		{"grpc.testingv3.SearchResponseV3.Result.Value.str", fdTestv3Byte},
 		{"grpc.testingv3.SearchResponseV3.State", fdTestv3Byte},
-		{"grpc.testingv3.SearchResponseV3.State.FRESH", fdTestv3Byte},
+		{"grpc.testingv3.SearchResponseV3.FRESH", fdTestv3Byte},
```
This was a bug in the previous formulation. In protobuf, enum values are defined at the same scope as their enclosing enum.

jhump commented on reflection/serverreflection.go:
```diff
@@ -65,291 +60,165 @@ type GRPCServer interface {
-		s.processEnum(fd, prefix, en)
+// NewServer returns a reflection server implementation using the given options.
+// It returns an error if the given options are invalid.
+func NewServer(opts ServerOptions) (rpb.ServerReflectionServer, error) {
```
I'm curious what your thoughts on this style of API, vs. `Register` where the server object is never exposed.

One advantage here is that it allows the caller to wrap/decorate the service. (Other formulation requires caller to use an interceptor to do so.)

Also, the options include an optional `GRPCServer`. But we'd _have_ to have a server to do the actual `Register...` call. So it feels like a confusing API to have a required server argument and then also an optional field in the options struct.

menghanl replied:
Or, we define an interface with `func GetServiceInfo()` (and you won't need the other field `ServiceNames []string`).

Did we talk about this interface option before? I don't remember if we got any conclusion.

jhump replied:
There is already an interface named `GRPCServer`. The list of `ServiceNames` is useful if the caller wants more control over exactly _which_ services are advertised by the reflection service.

menghanl replied:
Oh, right. I forgot we already made the `GRPCServer` interface.

What I meant is a smaller interface, with just `GetServiceInfo() map[string]grpc.ServiceInfo`.
You can still pass it the grpc server (that's the default behavior), or the dynamic factory.
Or if you want a static `[]string`, you can pass it an implementation that returns the list (and you won't need the other field).

jhump replied:
Okay, I think I see what you mean.

Would you object to me adding a utility function, like `ServiceInfoFromNames`, that takes a slice of names (`[]string`) and returns this new interface, that provides the service info?

jhump replied:
Done. (I went ahead and added the utility function I asked about. Let me know what you think.)

menghanl commented on reflection/serverreflection.go:
```diff
@@ -65,291 +60,165 @@ type GRPCServer interface {
+	// The list of service names. This should only be populated if Server is
+	// nil.
+	ServiceNames []string
+	// Optional resolver used to load descriptors.
+	DescriptorResolver protodesc.Resolver
+	// Optional resolver used to query for known extensions.
+	ExtensionResolver ExtensionResolver
```
Let's document the default value for these. `If unset, protoregistry.GlobalFiles will be used`

jhump replied:
These were documented above. So I rearranged a little and moved the relevant comments down to the fields' Go doc.

menghanl commented on reflection/serverreflection.go:
```diff
@@ -65,291 +60,165 @@ type GRPCServer interface {
+// NewServer returns a reflection server implementation using the given options.
+// It returns an error if the given options are invalid.
+func NewServer(opts ServerOptions) (rpb.ServerReflectionServer, error) {
+	if opts.Server != nil && len(opts.ServiceNames) > 0 {
```
(Related to the other comment, if we merge these two fields)
We can remove this error. Also clean up the code in `Register`.

jhump replied:
Done

jhump commented on reflection/serverreflection.go:
```diff
@@ -37,319 +37,194 @@ To register server reflection on a gRPC server:
-func (s *serverReflectionServer) processField(fd *dpb.FileDescriptorProto, prefix string, fld *dpb.FieldDescriptorProto) {
-	fldName := fqn(prefix, fld.GetName())
-	s.symbols[fldName] = fd
+// ServiceInfoFromNames returns a ServiceInfoProvider backed by the given
+// slice of service names. This allows for creating a reflection server
+// with exactly the given advertised services.
+func ServiceInfoFromNames(serviceNames []string) ServiceInfoProvider {
```
Let me know if you object to this helper. Though I think it's useful since it might not otherwise be obvious how to customize the set of advertised services.

menghanl replied:
I personally think this isn't necessary.

If your concern is that the interface is confusing (since only the map keys are used), we can change the interface to return `[]string`, and wrap `*grpc.Server` to implement the interface.

@dfawley WDYT

jhump replied:
Okay. I went ahead and removed it and added some extra documentation to the `Services` field in `ServerOptions`.

menghanl commented on reflection/serverreflection.go:
```diff
@@ -37,319 +37,194 @@ To register server reflection on a gRPC server:
-	for _, msg := range fd.MessageType {
-		s.processMessage(fd, prefix, msg)
+// NewServer returns a reflection server implementation using the given options.
+// It returns an error if the given options are invalid.
```
And to be a bit more cautious, mark the new functions/types as experimental?

jhump replied:
Done

menghanl commented on reflection/serverreflection.go:
```diff
@@ -37,319 +37,194 @@ To register server reflection on a gRPC server:
-func fileDescWithDependencies(fd *dpb.FileDescriptorProto, sentFileDescriptors map[string]bool) ([][]byte, error) {
-	r := [][]byte{}
-	queue := []*dpb.FileDescriptorProto{fd}
+func (s *serverReflectionServer) fileDescWithDependencies(fd protoreflect.FileDescriptor, sentFileDescriptors map[string]struct{}) ([][]byte, error) {
```
On `map[string]bool` vs `map[string]struct{}`.
1. I'm not sure which is more memory efficient, but I think the difference is small enough that we wouldn't care
2. In gRPC-Go, we've mostly (but not as an official policy) adopted `map[string]bool` as "set"s. Since if the entry doesn't exist, the map read returns the default bool (which is false). And we don't need to write `struct{}{}` when adding entries.

Let me know what you think.

jhump replied:
Done

menghanl reviewed (commented):
Overall LGTM.

jhump commented:
ping @dfawley

jhump commented on reflection/serverreflection.go:
```diff
@@ -37,314 +37,182 @@ To register server reflection on a gRPC server:
+	// This value will typically be a *grpc.Server. But the set of advertised
+	// services can be customized by wrapping a *grpc.Server or using an
+	// alternate implementation that returns a custom set of service names.
+	// It is okay for a custom implementation to return zero values for the
+	// grpc.ServiceInfo values in the map.
```
There are similar remarks there already:
https://github.com/grpc/grpc-go/pull/5197/files/a97ada894ee90340c4b53e0d0db7a7ec108e4283#diff-8aead0172d481e4bc1112c2b69e1e60cd91a32e638ff30f9905623ce41c47215R56-R57

jhump replied:
I'll remove this and put more clarification over in the other Go docs.

jhump commented on reflection/serverreflection.go:
```diff
@@ -37,314 +37,182 @@ To register server reflection on a gRPC server:
+//
+// The reflection service is only interested in the service names, but the
+// signature is this way so that *grpc.Server implements it.
+type ServiceInfoProvider interface {
```
Added here and to the `ExtensionResolver` interface, too.

jhump commented:
I re-ordered some elements in the file in my last commit, hoping to make the file layout a little more intuitive. All of the new experimental stuff now appears after the preferred non-experimental API elements.

I think I've addressed all comments. PTAL and let me know.

And thanks for the consideration/review!

dfawley reviewed (approved):
I like the code reorganization, thank you for the extra care.

jhump commented:
Thanks you, @dfawley and @menghanl!
